File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.java
import java.util.*;

class Solution {
    public String solution(String[] survey, int[] choices) {
        Map<Character, Integer> map = new HashMap<>();
        map.put('R', 0); map.put('T', 0);
        map.put('C', 0); map.put('F', 0);
        map.put('J', 0); map.put('M', 0);
        map.put('A', 0); map.put('N', 0);
        
        int[] score = {0, 3, 2, 1, 0, 1, 2, 3};
        
        String s; int c;
        for (int i = 0; i < survey.length; i++) {
            s = survey[i]; c = choices[i];
            if (c < 4) {
                map.put(s.charAt(0), map.get(s.charAt(0)) + score[c]);
            }
            if (c > 4) {
                map.put(s.charAt(1), map.get(s.charAt(1)) + score[c]);
            }
        }

        String[] type = {"RT", "CF", "JM", "AN"};
        String answer = "";
        for (int i = 0; i < 4; i++) {
            char a = type[i].charAt(0);
            char b = type[i].charAt(1);
            if (map.get(a) >= map.get(b)) {
                answer += a;
            } else {
                answer += b;
            }
        }
        return answer;
    }
}

File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.py
def solution(survey, choices):
    result = [{"R" : 0, "T" : 0}, {"C" : 0, "F" : 0}, {"J" : 0, "M" : 0}, {"A" : 0, "N" : 0}]
    number = {"RT" : 0, "TR" : 0, "CF" : 1, "FC" : 1, "JM" : 2, "MJ" : 2, "AN" : 3, "NA" : 3 }
    score = [0, 3, 2, 1, 0, 1, 2, 3]
    
    for i in range(len(survey)):
        c, s = choices[i], survey[i]
        idx = number[s]
        if c < 4:
            result[idx][s[0]] += score[c]
        if c > 4:
            result[idx][s[1]] += score[c]
    
    answer = ''
    for i in range(4):
        a, b = result[i].items()
        if a[1] >= b[1]:
            answer += a[0]
        else:
            answer += b[0]
    return answer